### ISSUE TITLE
TokenCache doesn't use `expires_on` value for access tokens

### ISSUE THREAD

**Issue by vbarbaresi** [AzureAD/microsoft-authentication-library-for-python#369]
**Describe the bug**
`TokenCache` ignores `expires_on` value for access tokens and recomputes the expiration from the current time instead.
This incorrect `expires_on` value is saved and returned by the cache, which leads to Azure Identity SDK considering expired tokens as valid.

**To Reproduce**
Simple script showing the issue:

```python
import msal


token_cache = msal.TokenCache()

expires_on = 1500000000

event = {
    "response": {"access_token": "foo", "expires_on": 1500000000},
    "scope": "scope",
}

token_cache.add(event=event)

cached_token = token_cache.find(msal.TokenCache.CredentialType.ACCESS_TOKEN)[0]

print("cached token expires_on:", cached_token["expires_on"])
print("expires_on was set to:", expires_on)
```
output is:
```
cached token expires_on: 1622627989
expires_on was set to: 1500000000
```

**Expected behavior**
We expect IMDS tokens cached from Azure Python SDK to honor the `expires_on` value from the access token instead of recomputing it from the current `time.time()`
https://github.com/Azure/azure-sdk-for-python/blob/azure-identity_1.6.0/sdk/identity/azure-identity/azure/identity/_internal/managed_identity_client.py#L88
Here you can see that Azure Identity client sets this `expires_on` but it is ignored in MSAL `TokenCache`

**What you see instead**
Paste the sample output, or add screenshots to help explain your problem.

**The MSAL Python version you are using**
1.12.0

**Additional context**
We are using our own caching layer for IMDS token requests. A token that is a few minutes old can be returned to our client.
As MSAL uses `now + expires_in` instead of the real `expires_on`, it can return expired tokens to Azure Identity with a wrong expiration date
Ultimately it leads to this authentication failure from Azure Python SDK

```
ErrorCode:AuthenticationFailed
Error:None
AuthenticationErrorDetail:Lifetime validation failed. The token is expired.
```

**Comment by rayluo** [comment]
So, you are trying to reverse-engineering MSAL Python's token cache to store tokens coming from a different source (in this case, tokens from IMDS). It is possible, but it needs to be done in standard-compliant way.

In the industry standard OAuth2, only `expires_in` is defined, so that is the valid *input* that MSAL Python operates with. The `expires_on` field , on the contrary, is invented by MSAL as an *output* field that stores the computed expiration time. You can not feed an output `expires_on` to MSAL Python and expects it to take it. It won't.

By the way, [the real world IMDS token response](https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http) would also contain `expires_in` (because IMDS protocol was designed to be compatible with OAuth2), so MSAL Python would still accept an IMDS token as long as it keeps that `expires_in` field. It is somewhat an unexpected side effect that the IMDS protocol also defines the `expires_on` which adds to the confusion.

In short, change your hack into using `expires_in`, and it would work.

```python
event = {
    "response": {"access_token": "foo", "expires_in": 3600},
    "scope": "scope",
}

token_cache.add(event=event)
```

**Comment by vbarbaresi** [comment]
I'm not reverse-engineering / hacking anything, I experienced this issue using azure-identity SDK.
azure-identity caches IMDS tokens using MSAL `TokenCache` here: https://github.com/Azure/azure-sdk-for-python/blob/azure-identity_1.6.0/sdk/identity/azure-identity/azure/identity/_internal/managed_identity_client.py#L85-L105

My code snippet was crafted just to reproduce what Azure SDK does, without requiring to use Azure SDK.

Indeed I saw that there is an `expire_in` field. But in our case, we are getting a cached "old" token from a separate caching layer. You can't expect this caching to adjust `expires_in` in real time: it's just a constant value set to `3600`

```python
event = {
    "response": {"access_token": "foo", "expires_in": 3600},
    "scope": "scope",
}

token_cache.add(event=event)
```

In this example, let's suppose `access_token` is 10 minutes old. It will expire in 50 minutes (`expires_on` would reflect that)
When you run `token_cache.add(event=event)`, `expires_on` will be recomputed from `now + expires_in` which will be 60 minutes in the future.
So Azure SDK will reuse this token for 60 minutes, even though it should expire in 50 minutes.

I can understand if this is a won't fix on MSAL side.
Though I find it weird that azure-identity is explicitly setting `expires_on` in the access token but it gets completely ignored by `TokenCache`.

**Comment by rayluo** [comment]
> In this example, let's suppose access_token is 10 minutes old. It will expire in 50 minutes (expires_on would reflect that)
When you run token_cache.add(event=event), expires_on will be recomputed from now + expires_in which will be 60 minutes in the future.

The specs-compliant way to achieve this is to have the incoming "event" to contain an adjusted `expires_in=50*60`, then everything will work as expected.

> So Azure SDK will reuse this token for 60 minutes, even though it should expire in 50 minutes.
>
> I can understand if this is a won't fix on MSAL side.
Though I find it weird that azure-identity is explicitly setting `expires_on` in the access token but it gets completely ignored by `TokenCache`.

Sounds like this should be addressed in azure-identity. Looping in its maintainer @chlowell for awareness.

**Comment by rayluo** [comment]
@vbarbaresi So I'm going to close this one as WontFix here in MSAL repo. But you may want to [start a new conversation in Azure Identity repo](https://github.com/Azure/azure-sdk-for-python/issues/new?assignees=&labels=&template=bug_report.md&title=).
